extensions/chromium/pdfHandler.js

```diff
@@ -18,183 +18,243 @@ limitations under the License.
+ *
+ * The logic here is the declarative version of the runtime logic in the
+ * webRequest.onHeadersReceived implementation at
+ * https://github.com/mozilla/pdf.js/blob/0676ea19cf17023ec8c2d6ad69a859c345c01dc1/extensions/chromium/pdfHandler.js#L34-L152
  */
-function isPdfFile(details) {
-  var header = getHeaderFromHeaders(details.responseHeaders, "content-type");
-  if (header) {
-    var headerValue = header.value.toLowerCase().split(";", 1)[0].trim();
-    if (headerValue === "application/pdf") {
-      return true;
-    }
-    if (headerValue === "application/octet-stream") {
-      if (details.url.toLowerCase().indexOf(".pdf") > 0) {
-        return true;
-      }
-      var cdHeader = getHeaderFromHeaders(
-        details.responseHeaders,
-        "content-disposition"
-      );
-      if (cdHeader && /\.pdf(["']|$)/i.test(cdHeader.value)) {
-        return true;
-      }
-    }
-  }
-  return false;
-}
+async function registerPdfRedirectRule() {
+  // "allow" means to ignore rules (from this extension) with lower priority.
+  const ACTION_IGNORE_OTHER_RULES = { type: "allow" };
 
-/**
- * Takes a set of headers, and set "Content-Disposition: attachment".
- * @param {Object} details First argument of the webRequest.onHeadersReceived
- *                         event. The property "responseHeaders" is read and
- *                         modified if needed.
- * @returns {Object|undefined} The return value for the onHeadersReceived event.
- *                             Object with key "responseHeaders" if the headers
- *                             have been modified, undefined otherwise.
- */
-function getHeadersWithContentDispositionAttachment(details) {
-  var headers = details.responseHeaders;
-  var cdHeader = getHeaderFromHeaders(headers, "content-disposition");
-  if (!cdHeader) {
-    cdHeader = { name: "Content-Disposition" };
-    headers.push(cdHeader);
+  // Redirect to viewer. The rule condition is expected to specify regexFilter
+  // that matches the full request URL.
+  const ACTION_REDIRECT_TO_VIEWER = {
+    type: "redirect",
+    redirect: {
+      // DNR does not support transformations such as encodeURIComponent on the
+      // match, so we just concatenate the URL as is without modifications.
+      // TODO: use "?file=\\0" when DNR supports transformations as proposed at
+      // https://github.com/w3c/webextensions/issues/636#issuecomment-2165978322
+      regexSubstitution: VIEWER_URL + "?DNR:\\0",
+    },
+  };
+
+  // Rules in order of prority (highest priority rule first).
+  // The required "id" fields will be auto-generated later.
+  const addRules = [
+    {
+      // Do not redirect for URLs containing pdfjs.action=download.
+      condition: {
+        urlFilter: "pdfjs.action=download",
+        resourceTypes: ["main_frame", "sub_frame"],
+      },
+      action: ACTION_IGNORE_OTHER_RULES,
+    },
+    {
+      // Do not redirect for URLs containing pdfjs.action=download.
```

Comment:
This comment doesn't seem to match the condition below; is it perhaps accidentally copy/pasted from the one above?